apiserver/remotefirewaller/ingressaddresswatcher.go

```diff
@@ -0,0 +1,166 @@
+	w := &IngressAddressWatcher{
+		backend: backend,
+		appName: appName,
+		rel:     rel,
+		known:   make(map[string]string),
+		out:     make(chan []string),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, err
+}
+
+func (w *IngressAddressWatcher) loop() error {
+	ruw, err := w.rel.WatchUnits(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(ruw); err != nil {
+		return errors.Trace(err)
+	}
+
+	var (
+		sentInitial bool
+		out         chan<- []string
+	)
+	// addresses holds the current set of known addresses.
+	addresses := make(set.Strings)
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case c, ok := <-ruw.Changes():
+			if !ok {
+				return watcher.EnsureErr(ruw)
+			}
+			// A unit has entered or left scope.
+			// Get the new set of addresses resulting from that
+			// change, and if different to what we know, send the change.
+			merged, err := w.mergeAddresses(addresses, c)
+			if err != nil {
+				return err
+			}
+			changed := !merged.Difference(addresses).IsEmpty() || !addresses.Difference(merged).IsEmpty()
+			if !sentInitial || changed {
+				addresses = merged
+				out = w.out
+			} else {
+				out = nil
+			}
+		case out <- formatAsCIDR(addresses.Values()):
+			sentInitial = true
+			out = nil
+		}
+	}
+}
+
+func formatAsCIDR(addresses []string) []string {
+	result := make([]string, len(addresses))
+	for i, a := range addresses {
+		result[i] = a + "/32"
+	}
+	return result
+}
+
+func (w *IngressAddressWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+func (w *IngressAddressWatcher) mergeAddresses(addresses set.Strings, c params.RelationUnitsChange) (set.Strings, error) {
+	result := set.NewStrings(addresses.Values()...)
+	for name := range c.Changed {
+
+		u, err := w.backend.Unit(name)
+		if errors.IsNotFound(err) {
+			logger.Warningf("unit %s is not found, can't get address", name)
+			continue
+		}
+		if err != nil {
+			return result, err
+		}
+
+		// TODO - start watcher to pick up machine address changes
+		// We need to know whether to look at the public or cloud local address.
+		// For now, we'll use the public address and later if needed use a watcher
+		// parameter to look at the cloud local address.
+		addr, err := u.PublicAddress()
+		if errors.IsNotAssigned(err) {
+			logger.Warningf("unit %s is not assigned to a machine, can't get address", name)
+			continue
+		}
+		if network.IsNoAddressError(err) {
+			logger.Warningf("unit %s has no public address", name)
+			continue
+		}
+		if err != nil {
+			return result, err
+		}
+		logger.Debugf("unit %q has public address %q", name, addr.Value)
+		result.Add(addr.Value)
+		w.known[name] = addr.Value
+	}
+	for _, name := range c.Departed {
+		// If the unit is departing and we have seen it's address,
+		// remove the address.
+		address, ok := w.known[name]
+		if ok {
+			delete(w.known, name)
+			result.Remove(address)
```

Comment:
this assumes that there's 1:1 between RUW and address, which is not necessarily true. you could have two units on the same machine. so I think you need a counter as well.